### PR TITLE
Feature: Make non-zero exit code available when compiler compile fails

### LINKED ISSUE #6: Module does not provide an exit code when compiler compile fails
## Bug

cmdlet does not exit with non-zero exit code when compiler compile fails.

## Expectation

cmdlet does exit with non-zero exit code when compiler compile fails.

Two ways to get a non-zero exit code in powershell:

1. `throw` a terminating error in the script (we lose the exit code of the compiler)
2. `exit 1` in the script (we exit the entire powershell process)
3. Populate the `$global:LASTEXITCODE` for consumer to read.

Edit: I think its better to pass exit code via option 3 instead of the first two options.

### PR DESCRIPTION
**Pull request by leojonathanoh**
Fixes #6

### CODE REVIEW

**Comment by leojonathanoh** [comment]
For windows, seems like `compile.exe` is not producing a non-zero exit code: See [job](https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=282&view=logs&j=a4c7e33e-604c-560c-49b1-4d6828efc661&t=3de151fa-1c49-59eb-ac4b-c29a43df6985&l=766) and [job](https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=282&view=logs&j=a4c7e33e-604c-560c-49b1-4d6828efc661&t=3de151fa-1c49-59eb-ac4b-c29a43df6985&l=880)

For *nix, Need to find out what went wrong here: [job](https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=282&view=logs&j=1db8b590-9a39-57c2-14e8-7f7a62d6f140&t=2dedac55-9a9b-558f-b684-a6f2168a0c85&l=880)

**Comment by leojonathanoh** [comment]
There's some failed checks.

As discussed in https://github.com/startersclan/Compile-SourceScript/pull/2#issuecomment-623267891, if we decide to no longer support wrappers `compile.exe` and `compile.sh` which do not provide proper exit codes, then in this PR it will be much easier to fix existing failed checks.

**Comment by joeltimothyoh** [comment]
> 
> 
> There's some failed checks.
> 
> As discussed in [#2 (comment)](https://github.com/startersclan/Compile-SourceScript/pull/2#issuecomment-623267891), if we decide to no longer support wrappers `compile.exe` and `compile.sh` which do not provide proper exit codes, then in this PR it will be much easier to fix existing failed checks.

I'd actually be open adding support for exit codes for when compiler wrapper scripts are invoked, and more so than supporting #2. Should we decide to continue support for invoking compiler wrappers, I'll see how I can contribute to make this feature robust if possible.

**Review by joeltimothyoh — COMMENTED** [review]
lgtm. Please review @leojonathanoh

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 [hunk 1]
```diff
@@ -145,11 +146,48 @@ function Compile-SourceScript {
                     $sourceFile.Name
                 )
             }
+            $returnExitCode = $false
```
Comment (leojonathanoh) [inline src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: Could this variable be named something more understandable? The cmdlet `Compile-Sourcescript` doesn't actually return an exit code, but populates a global `$global:LASTEXITCODE`.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 [hunk 2]
```diff
@@ -145,11 +146,48 @@ function Compile-SourceScript {
...
+            }elseif ($MOD_NAME -eq 'amxmodx') {
+                $pluginErrorRegexPattern = '^.*\.sma\(\d+\)\s*:\s*error (\d+)'
+                $returnExitCode = $true
+            }
             New-Item -Path $COMPILED_DIR -ItemType Directory -Force | Out-Null
 
             # Begin compilation
             if ($PSBoundParameters['SkipWrapper']) { "Compiling $($sourceFile.Name)..." | Write-Host -ForegroundColor Yellow }
-            Start-Process @processArgs
+
```
Comment (leojonathanoh) [inline src/Compile-SourceScript/Public/Compile-SourceScript.ps1:168]: This block of code could be made into a separate function called `Compile-Plugin` which returns the exit code, so that the flag `$returnExitCode` is not needed. Instead, the global `$global:LASTEXITCODE` population is done right after `Compile-Plugin` function is called.

File: test/scripts/integration/Run-IntegrationTests.ps1 [hunk 3]
```diff
@@ -34,8 +38,17 @@ $cmdArgs = @{
...
 & $functionTestScriptBlock
 
+"`n[sourcemod] Compile plugin via wrapper, bad plugin" | Write-Host
+$cmdArgs = @{
+    File = "$PSScriptRoot\..\..\mod\sourcemod\addons\sourcemod\scripting\plugin1_bad.sp"
+    Force = $true
+}
+$iterations = 2
+$expectedExitCode = 1
+& $functionTestScriptBlock
```
Comment (leojonathanoh) [inline test/scripts/integration/Run-IntegrationTests.ps1:51]: `&functionTestScriptBlock` could be made into a function called `Test-CompilePlugin`, so that scoping and readability is better.

File: test/scripts/integration/Run-IntegrationTests.ps1 [hunk 4]
```diff
@@ -60,6 +83,16 @@ if ($env:OS -eq 'Windows_NT') {
         File = "$PSScriptRoot\..\..\mod\amxmodx\addons\amxmodx\scripting\plugin1.sma"
         Force = $true
     }
+    $expectedExitCode = 0
+    $iterations = 2
+    & $functionTestScriptBlock
+
+    "`n[amxmodx] Compile plugin via wrapper, bad plugin" | Write-Host
```
Comment (leojonathanoh) [inline test/scripts/integration/Run-IntegrationTests.ps1:90]: The exit codes produced by `compile.exe` and `compile.sh` might be wrong when multiple  `.sp` or `.sma` are passed to it. See https://github.com/startersclan/Compile-SourceScript/pull/2#issuecomment-623314082. Hence, when testing a compiler wrapper, more than one good test file and one bad test file should be passed to it, with test cases containing a different order combination of its tested files(i.e. one test case with good followed by bad, and another test case with bad followed by good, where both test cases should expect a non-zero exit code.)

**Comment by joeltimothyoh** [comment]
Pinging @leojonathanoh

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 1 above)
Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: I'm not sure of a good name. I think it was named with the assumption that the exit code would be returned by the `Compile-SourceScript` instead of the compiler. Albeit as you mention, in both cases `$global:LASTEXITCODE` is populated, and there being no reason for naming the variable as such. Perhaps make a suggestion for what the variable name should be, since the one I came up with seem senseless.

Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: its related to https://github.com/startersclan/Compile-SourceScript/pull/7#discussion_r464560525. Then we can mark this as resolved.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 2 above)
Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:168]: Would you like to make those changes you have in mind? I see splitting up to be necessary only if the function is getting too polluted. The function `Compile-SourceScript` however does lots of other things that could possibly be moved to separate functions. Whether it's necessary to be splitting up the code into multiple functions is a question, though I think I've previously mentioned how PowerShell cmdlets themselves make it possible to perform multiple steps in a short amount of code. Splitting up code when the same can be achieved via the in-built cmdlets themselves is a common mistake I've made in the past.

File: test/scripts/integration/Run-IntegrationTests.ps1 (hunk 4 above)
Comment (joeltimothyoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:90]: I believe they would be right based on the regex patterns in https://github.com/startersclan/Compile-SourceScript/pull/7/files#diff-1ff1c12c3ce44f26b1b450870ac4db81R151-R156. Unless you're talking about passing to the compilers non `.sp` or `.sma` files, in which case is taken care of in https://github.com/startersclan/Compile-SourceScript/pull/7/files#diff-1ff1c12c3ce44f26b1b450870ac4db81R53-R57.

Please review the build logs to see how it now works on Window and LInux.

File: test/scripts/integration/Run-IntegrationTests.ps1 (hunk 3 above)
Comment (joeltimothyoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:51]: The scriptblock [$functionTestScriptBlock](https://github.com/startersclan/Compile-SourceScript/pull/7/files#diff-56c3cff6bb7f2d2d84c4f3d644a5fb0aR9-R39) is simply a reusable piece of code that can be used in integration tests for testing PowerShell modules. It's just something I came up that can be used across multiple repos which isn't overly complicated I feel. Functions could be used too, though scoping of variables cannot be used in the same way as is used now with `script:` etc that makes it possible to prepare multiple parameters necessary for calling the tested function. Feel free to come up with one if you have the time to; but it had better be simple and as robust as it is now.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 [hunk 5]
```diff
@@ -145,11 +146,48 @@ function Compile-SourceScript {
...
+                $tempDir = Join-Path ([System.IO.Path]::GetTempPath()) (New-Guid).Guid
+                $stdoutFile = Join-Path $tempDir 'stdout'
+                $stderrFile = Join-Path $tempDir 'stderr'
+                $processArgs['RedirectStandardOutput'] = $stdoutFile
+                $processArgs['RedirectStandardError'] = $stderrFile
+                New-Item $tempDir -ItemType Directory -Force > $null
+                $p = Start-Process @processArgs
+                $stdout = Get-Content $stdoutFile
+                $stdout | Write-Host
+                $stderr = Get-Content $stderrFile
```
Comment (joeltimothyoh) [inline src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]: The variable `$stderr` is never used. Based on my testing the variable is always empty together with the stderr file. I believe the compiler and compiler wrappers only use stdout. In which case it may be better to not use redirection to stdout and stderr files, since redirection also means the inability to display output in the console in real-time and in the correct order. I added outputting of stdout as a workaround, but adding outputting of stderr would be hackerish leading to incorrect order of corresponding stdout and stderr lines for each instance of compilation as in the case when multiple scripts are compiled when the compiler wrappers are used. Because of this reason I think it would be better if redirection of stdout and stderr to files aren't used, which would be possible if we are confident that the compilers and compiler wrappers only output to stdout.

Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]: We can use `Tee-Object $stdoutFile | Write-Host` write to file and print to console @joeltimothyoh

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 2 above)
Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:168]: Yes, the cmdlet is getting very polluted. Better to use a more functional approach, so as to be able to add Pester unit tests which hasn't been implemented yet. I can make the necessary changes

File: test/scripts/integration/Run-IntegrationTests.ps1 (hunk 3 above)
Comment (leojonathanoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:51]: @joeltimothyoh I recommend using Pester integration tests, but in a separate MR, instead of our own integration test implementation. So as to keep changes specific to the MR.

Comment (joeltimothyoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:51]: Not a fan of an test-driven approach. I’m unsure I’ll actually bother writing such tests for a simple tool like this. Yes we can always open another PR for those changes.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 5 above)
Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]: That would mean not using redirect params for `Start-Process`. Tee-Object is perfect for if we’re confident stdout is only used by compiler and compiler wrappers as I mentioned.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 2 above)
Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:168]: The function you mean. I think it has vastly increased in complication due to the use of redirect parameters of `Start-Process` and generation of temporary files. The function was acceptable without the changes from this PR. If we revised to not use redirect output to temporary files as discussed in https://github.com/startersclan/Compile-SourceScript/pull/7#discussion_r468039125, it would save us the hassle of refactoring.

Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:168]: A module's main function is called a cmdlet, while it's private functions are called functions. I meant the cmdlet as in `Compile-SourceScript`. 

> The function was acceptable without the changes from this PR. If we revised to not use redirect output to temporary files as discussed in #7 (comment), it would save us the hassle of refactoring.

I'm adding in and you will get what I mean

File: test/scripts/integration/Run-IntegrationTests.ps1 (hunk 3 above)
Comment (leojonathanoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:51]: @joeltimothyoh it's not about test-driven, but more about taking a modular approach that results in more testable units of code.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 5 above)
Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]: @joeltimothyoh based on the logs, I don't see any stderr output emitted by the compiler or the compiler wrapper. I think we can be confident to rely solely on regex patterns on the stdout, which will work consistently in detecting compilation errors regardless of whether compilation is done via compiler or the compiler wrapper, and not rely on exit codes from the process.

Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]: EDIT: second link was wrong. Updated.

To capture `stdout` of `Start-Process` is not as easy as `Start-Process .. | Tee-Object`. It's tricky. See https://stackoverflow.com/questions/8761888/capturing-standard-out-and-error-with-start-process and https://adamtheautomator.com/powershell-start-process/. I suggest we stick with `Start-Process -RedirectStandardOutput x -RedirectStandardError y`. @joeltimothyoh I'm pushing changes

Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]:
```powershell
$p = start-process -FilePath powershell -ArgumentList '-c', 'foreach ($i in (1..5)){ sleep 1; echo hi; }' -NoNewWindow  -Wait -PassThru -RedirectStandardOutput ~/stdout.log -RedirectStandardError ~/stderr.log
```

Once `-RedirectStandardOutput` and `-RedirectStandardError y` are used then `Start-Process` will not print `stdout` and `stderror` to the console. So that might be a little impairing for the developer who might want to see realtime compilation progress.

Since compilation is quick it should not take too long before compilation would be completed and printed to console, so I think this should be fine for the developer.

Thoughts @joeltimothyoh

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 2 above)
Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:168]: > A module's main function is called a cmdlet, while it's private functions are called functions. I meant the cmdlet as in Compile-SourceScript.

Not true, it depends on whether functions are exported as a cmdlet or not:

- https://github.com/startersclan/Compile-SourceScript/blob/9d4e5de4e9ffd21b58b34580366a3a88d91bc742/build/definitions/modulemanifest/definition.ps1#L26-L29

```
Import-Module .\src\Compile-SourceScript\Compile-SourceScript.psm1 -Force
$ Get-Module Compile-SourceScript

ModuleType Version    Name                                ExportedCommands    
---------- -------    ----                                ----------------    
Script     0.0        Compile-SourceScript                Compile-SourceScript

$ Get-Command Compile-SourceScript

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Function        Compile-SourceScript                               0.0        Compile-SourceScript
```

Anyway, that's besides the point.

Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:168]: @joeltimothyoh at least from my understanding it's a common convention that a module's main function(s) should be exported as Cmdlet

File: test/scripts/integration/Run-IntegrationTests.ps1 (hunk 3 above)
Comment (joeltimothyoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:51]: I'd steer clear from writing unit tests for something so fragile such as what's worked on in this PR. If functions are written from a low-level, it'd then be fine to have unit tests, but for changes such as those involved in this PR, there's much that we *don't* know and which involves invocation of multiple compilers and compiler wrappers, in all honesty, it's better to leave out the unit tests unless behavior of said scenarios has shown to be predictable and design of the module and its functions are mostly set in place.

For simplicity I think it's better to aim for real world uses first, and those simply involve integration tests. The module would not only be simpler to maintain, but can be refactored or revised accordingly quickly to fit the most practical of use cases.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 5 above)
Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]: > Once -RedirectStandardOutput and -RedirectStandardError y are used then Start-Process will not print stdout and stderror to the console. So that might be a little impairing for the developer who might want to see realtime compilation progress.

This is the very issue I've already explained above in https://github.com/startersclan/Compile-SourceScript/pull/7#discussion_r468039125 and https://github.com/startersclan/Compile-SourceScript/pull/7#discussion_r468095470. It's better not to redirect any output to files with `Start-Process`. Instead, we should treat the console output as the first class citizen, which can be redirected using Tee-Object to file(s) or variables with `Start-Process`. It would be even better if `Tee-Object` could be done without; the fewer the moving parts involved in compilation processes, the better.

File: test/scripts/integration/Run-IntegrationTests.ps1 (hunk 3 above)
Comment (leojonathanoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:51]: @joeltimothyoh sure. in the long run, for better ci processes esp since this is open source, having unit tests helps speed up contributors who can at least test majority of the code on their machine. Just a suggestion that could be implemented in the future

File: test/scripts/integration/Run-IntegrationTests.ps1 (hunk 4 above)
Comment (leojonathanoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:90]: now that we are using regex patterns as discussed in https://github.com/startersclan/Compile-SourceScript/pull/7#discussion_r468210203, failing to detect errors in compilation via compilation wrappers should no longer happen.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 2 above)
Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:168]: now that we are using regex patterns as discussed in #7 (comment), there is no more need for this flag (which served to inject a non-zero exit code into `global:LASTEXITCODE` for cases where the compiler binary or compiler wrapper do not return a non-zero exit code even when compilation errors occured). I removed the flag in https://github.com/startersclan/Compile-SourceScript/pull/7/commits/5f4dcebccfd11f2520a3ffefb9caad291815487d

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 1 above)
Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: Resolved in https://github.com/startersclan/Compile-SourceScript/pull/7#discussion_r468257270

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 5 above)
Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]: @joeltimothyoh the problem is that there is no way of capturing the stdout returned by `Start-Process`, as i explained above, see https://github.com/startersclan/Compile-SourceScript/pull/7#discussion_r468238198. 

I agree the best is to preserve console output, but it's not possible without using `Invoke-Process` as suggested in the second link.

File: test/scripts/integration/Run-IntegrationTests.ps1 (hunk 4 above)
Comment (joeltimothyoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:90]: That would still be an assumption. I don't know the compilers well enough to say if that regex pattern suffices in capturing all possible errors that could occur. In fact I think what's more robust is revising the regex pattern to instead account for the pattern such as:

- sourcemod - https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=319&view=logs&j=cf913be4-6704-58f5-cde6-258cc8aef7bf&t=c3dd6012-84fa-5b88-85db-ffd0f16e84df&l=545 instead of https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=319&view=logs&j=cf913be4-6704-58f5-cde6-258cc8aef7bf&t=c3dd6012-84fa-5b88-85db-ffd0f16e84df&l=545
- amxmodx - https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=319&view=logs&j=cf913be4-6704-58f5-cde6-258cc8aef7bf&t=c3dd6012-84fa-5b88-85db-ffd0f16e84df&l=887 instead of https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=319&view=logs&j=cf913be4-6704-58f5-cde6-258cc8aef7bf&t=c3dd6012-84fa-5b88-85db-ffd0f16e84df&l=885

Comment (leojonathanoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:90]: > That would still be an assumption. I don't know the compilers well enough to say if that regex pattern suffices in capturing all possible errors that could occur. In fact I think what's more robust is revising the regex pattern to instead account for the pattern such as:

> - `sourcemod` - https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=319&view=logs&j=cf913be4-6704-58f5-cde6-258cc8aef7bf&t=c3dd6012-84fa-5b88-85db-ffd0f16e84df&l=549 instead of https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=319&view=logs&j=cf913be4-6704-58f5-cde6-258cc8aef7bf&t=c3dd6012-84fa-5b88-85db-ffd0f16e84df&l=545

We could match several lines. E.g. regex `^Compilation aborted\.` and `^\d+\s+Errors?\.`

```
plugin1_bad.sp(283) : error 029: invalid expression, assumed zero
plugin1_bad.sp(283) : fatal error 190: too many error messages on one line

Compilation aborted.
4 Errors.
```

> amxmodx - https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=319&view=logs&j=cf913be4-6704-58f5-cde6-258cc8aef7bf&t=c3dd6012-84fa-5b88-85db-ffd0f16e84df&l=887 instead of https://dev.azure.com/startersclan/Compile-SourceScript/_build/results?buildId=319&view=logs&j=cf913be4-6704-58f5-cde6-258cc8aef7bf&t=c3dd6012-84fa-5b88-85db-ffd0f16e84df&l=885

We could match several lines. E.g. regex `(compile failed\)\.$` and `^\d+\s+Errors?\.`

```
plugin2_bad.sma(6) : error 010: invalid function or declaration

1 Error.
Could not locate output file compiled/plugin2_bad.amx (compile failed).
End of compilation.
```

I'll add those to the regex patterns if its ok @joeltimothyoh

Comment (joeltimothyoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:90]: I think those patterns would definitely do better. @leojonathanoh

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 5 above)
Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]: > @joeltimothyoh the problem is that there is no way of capturing the stdout returned by Start-Process, as i explained above, see #7 (comment).

I'll have to debug to find out.

> I agree the best is to preserve console output, but it's not possible without using Invoke-Process as suggested in the second link.

`Invoke-Process` as in https://www.powershellgallery.com/packages/Invoke-Process/1.4? If possible it'd be better to stick to in-built cmdlets.

File: test/scripts/integration/Run-IntegrationTests.ps1 (hunk 4 above)
Comment (leojonathanoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:90]: done

Comment (leojonathanoh) [reply test/scripts/integration/Run-IntegrationTests.ps1:90]: Fixed in https://github.com/startersclan/Compile-SourceScript/pull/7/commits/d9751aaa570f1164bfdc362d6f5c57fcdf2cc2b4

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 5 above)
Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]: > @joeltimothyoh the problem is that there is no way of capturing the stdout returned by Start-Process, as i explained above, see #7 (comment).

I've tested and this seems to be correct.

> Since compilation is quick it should not take too long before compilation would be completed and printed to console, so I think this should be fine for the developer.

I think so too. The order of the output should be fine so long as the compiler doesn't output to stderr.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 [hunk 6]
```diff
@@ -123,33 +128,60 @@ function Compile-SourceScript {
...
             if ($PSBoundParameters['SkipWrapper']) { "Compiling $($sourceFile.Name)..." | Write-Host -ForegroundColor Yellow }
-            Start-Process @processArgs
+
+            # Compile
+            $global:LASTEXITCODE = 0
+            $p = Start-Process @processArgs
+            $stdout = Get-Content $stdoutFile
+            $stdout | Write-Host
+            $stderr = Get-Content $stderrFile
+            foreach ($line in $stdout) {
```
Comment (joeltimothyoh) [inline src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: Is there a need for outputting stderr?

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 [hunk 7]
```diff
@@ -123,33 +128,60 @@ function Compile-SourceScript {
...
+                if ($line -match $MOD[$MOD_NAME]['compiler'][$OS]['error_regex']) {
+                    $global:LASTEXITCODE = 1
+                    break
+                }
+            }
+
+            # Cleanup
+            Remove-Item $stdInFile -Force
+            Remove-Item $tempDir -Recurse -Force
```
Comment (joeltimothyoh) [inline src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: These would not execute if the compile process terminates prematurely.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 6 above)
Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: As we said, `stderr` will be empty. No harm in outputting, though if it were not, because `stdout` and `stderr` are separated, the output will no longer be in the original sequence if printing `stderr` after printing `stdout`.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 7 above)
Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: Should not have a terminating at compile, or else use a `try catch` block and `finally` here to cleanup.

Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: Yes, a nested try-catch-finally block should work.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 6 above)
Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: We could add it in then.

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 7 above)
Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: or we could ensure `ErrorActionPreference = ‘Continue’` for that part

Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: Which part?

Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: when preparing and compiling and cleaning up. same block of code try catch would wrap around. @joeltimothyoh

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 6 above)
Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: Should we take the risk to output stderr or stay confident about just outputting stdout only?

Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: i dont see any risk with outputting `stderr`, except the scenario where the lack of continuity in the output does not match the developer's expectation. 

A benefit in outputting `stdout` is in the case where there are missing messages in `stdout`, with `stderr` a developer might be able to at least be able to have some sort of output to fix any issues which cause the compile to spit out messages on `stderr` which we currently have no knowledge of. 

@joeltimothyoh

Comment (joeltimothyoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: Maybe you could commit the changes you think makes most sense so we can close this PR? @leojonathanoh

Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: fixed

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 5 above)
Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1:174]: Fixed. See related discussion in https://github.com/startersclan/Compile-SourceScript/pull/7/files/eff914f808b9f4883e8d064020bdc8422ac214e8#r479858193

File: src/Compile-SourceScript/Public/Compile-SourceScript.ps1 (hunk 7 above)
Comment (leojonathanoh) [reply src/Compile-SourceScript/Public/Compile-SourceScript.ps1]: Fixed.

**Review by joeltimothyoh — APPROVED** [review]
lgtm

**Comment by joeltimothyoh** [comment]
Merging.

**Comment by joeltimothyoh** [comment]
Should we do a merge commit or squash merge? @leojonathanoh

**Comment by leojonathanoh** [comment]
either will do

**Comment by joeltimothyoh** [comment]
In terms of OSS workflow and git history? @leojonathanoh

**Comment by joeltimothyoh** [comment]
Pinging @leojonathanoh

**Comment by leojonathanoh** [comment]
i prefer merge commits since it’s more traceable

**Comment by joeltimothyoh** [comment]
Ok but we’ve already begun using squash commits in some other projects. Also release notes would be less readable or understandable. @leojonathanoh

**Comment by joeltimothyoh** [comment]
Merging via a merge commit as discussed with @leojonathanoh.
